File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/math/P398_RandomPickIndex.py
import random
from typing import List


class Solution:

    def __init__(self, nums: List[int]):
        self.nums = nums

    def pick(self, target: int) -> int:
        count, res = 0, 0
        for i, num in enumerate(self.nums):
            if num == target:
                count += 1
                if random.randrange(count) == 0:
                    res = i
        return res

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/math/P398_RandomPickIndex.java
package com.dsa.leetcode.labuladong.math;

import java.util.*;

/**
 * @since 2024/7/27 16:53
 */
public class P398_RandomPickIndex {

    class Solution {
        Random random;
        Map<Integer, List<Integer>> map;

        public Solution(int[] nums) {
            random = new Random();
            map = new HashMap<>();
            for (int i = 0, len = nums.length; i < len; i++) {
                map.putIfAbsent(nums[i], new ArrayList<>());
                map.get(nums[i]).add(i);
            }
        }

        public int pick(int target) {
            List<Integer> integers = map.get(target);
            return integers.get(random.nextInt(integers.size()));
        }
    }
}
